src/librustc_middle/ty/sty.rs

```diff
@@ -2178,21 +2180,38 @@ impl<'tcx> TyS<'tcx> {
 }
 
 /// Typed constant value.
-#[derive(Copy, Clone, Debug, Hash, RustcEncodable, RustcDecodable, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Debug, Hash, RustcEncodable, RustcDecodable, Eq, Ord, PartialOrd)]
 #[derive(HashStable)]
 pub struct Const<'tcx> {
     pub ty: Ty<'tcx>,
 
     pub val: ConstKind<'tcx>,
+
+    /// We compare `Const` by pointer equality. This would be incorrect
+    /// if we were able to create Const values on the stack.
+    ///
+    /// Only allow the creation of consts using `tcx.mk_const(ty, val)`.
+    pub _priv: PrivateMarker,
```

Comment:
Why is this `pub`?

Reply:
> This field has to be public, to allow pattern matching on consts

Not completely sure if this is the best approach :thinking:

Reply:
Private fields don't prevent pattern matching, you can still do this:
```rust
Const {
  ty,
  val,
  ..
}
```

Reply:
I thought I tried that and remembered that I had a similar problem with `#[non_exhaustive]` in the past :sweat_smile: 

Well, looks like I should clean this up then.
Thanks x.x